Add pass event

danielimmke opened the pull request:
#### Checklist

* [x] Use a separate branch in your local repo (not `master`).
* [x] Test coverage is 100% (or you have a story for why it's ok).

For issue #477 - I did my best based on my understanding of the codebase. It appears to work well. I understand there's probably going to be some changes needed before it can be moved in but I am happy to make some adjustments if needed.

nicolodavis commented on src/core/flow.js:
```diff
@@ -529,6 +529,19 @@ export function Flow({ moves, phases, endIf, turn, events, plugins }) {
+    if (arg && true === arg.remove) {
+      const playOrder = [...ctx.playOrder];
+
+      playOrder.splice(playOrder.indexOf(ctx.currentPlayer), 1);
```
This can be more efficiently written as:
```
const playOrder = playOrder.filter(i => i != ctx.currentPlayer);
ctx = { ...ctx, playOrder };
```

nicolodavis replied:
However, you should actually be using `playerID` (the player that called the event) rather than `currentPlayer`.  This value is passed around to all events and is available if you just extract it (similar to `EndStage` on line 563).

nicolodavis commented on src/core/flow.js:
```diff
@@ -529,6 +529,19 @@ export function Flow({ moves, phases, endIf, turn, events, plugins }) {
+      ctx = { ...ctx, playOrder };
+
+      if (0 === playOrder.length) {
+        EndPhaseEvent(state);
```
`state = EndPhaseEvent(state)`

danielimmke commented:
Hey Nicolo, thank you for the very quick review time! I have a new commit with the changes you requested please let me know what you think.

nicolodavis commented:
Looks good!  Can you add some unit tests in `flow.test.js` demonstrating the usage of the Pass event?

danielimmke commented:
Sure thing, I will try to have that this week.

danielimmke commented:
Hey Nicolo,

So I basically copied over the tests for endTurn and removed some items to prevent redundancy (since `pass` calls the endTurn method under the hood and added a specific test for the `remove` parameter. Seems to work!

nicolodavis commented on src/core/flow.js:
```diff
@@ -529,6 +529,18 @@ export function Flow({ moves, phases, endIf, turn, events, plugins }) {
+      ctx = { ...ctx, playOrder };
+
+      if (0 === playOrder.length) {
+        state = EndPhaseEvent(state);
```
Could you add a test that also captures ending the phase?

danielimmke commented:
Ok - I got that fixed. I think I broke my personal record for "most amount of time spent debugging for something that turned out to be a one line code change" - lol.

I also added an additional test to check that scenario.

nicolodavis commented on src/core/flow.js:
```diff
@@ -529,6 +529,18 @@ export function Flow({ moves, phases, endIf, turn, events, plugins }) {
+      ctx = { ...ctx, playOrder };
+
+      if (0 === playOrder.length) {
+        return EndPhaseEvent(state);
```
Instead of calling `EndPhaseEvent` directly here, you should queue it for processing by adding it to `next`.  See line 377 for an example of how to do this.

nicolodavis commented on src/core/turn-order.js:
```diff
@@ -259,12 +259,23 @@ export function UpdateTurnOrderState(G, ctx, turn, endTurnArg) {
+    Object.keys(endTurnArg).forEach(arg => {
+      switch (arg) {
+        case 'remove':
+          currentPlayer = getCurrentPlayer(ctx.playOrder, playOrderPos) || '0';
```
Would this ever fallback on the `0` at the end of the line?

danielimmke replied:
I was basically attempting to do what adjusting playOrderPos in the endTurn method would do (catch out of bounds.) Adjusting that makes this unnecessary I think.

nicolodavis commented on src/core/flow.js:
```diff
@@ -529,6 +529,18 @@ export function Flow({ moves, phases, endIf, turn, events, plugins }) {
+    playerID = playerID || state.ctx.currentPlayer;
+
+    if (arg && true === arg.remove) {
+      const playOrder = [...ctx.playOrder].filter(i => i != playerID);
```
If you don't also adjust `playOrderPos` here you might leave it out of bounds (when removing the player at the end of `playOrder`).

The following test demonstrates the problem:

```js
  test('test', () => {
    let t = state;
    t = flow.processEvent(t, gameEvent('pass'));
    t = flow.processEvent(t, gameEvent('pass'));
    t = flow.processEvent(t, gameEvent('pass', { remove: true }));
    expect(t.ctx.currentPlayer).toBe('0');   // this fails because currentPlayer is undefined here.
  });
```

danielimmke replied:
Good idea. Added my idea about it and added that test as well for posterity. Thanks for the review.

nicolodavis commented:
Nice job @danielimmke!

Mohre commented:
Hi. Where can i find the documentation of this feature?
